gate[] sorted_gate_array
